### PR TITLE
Feature: API Service

### PR DESCRIPTION
**Pull request by YehorPytomets**
### Changes introduced by this PR:
* Add `api service`. 
* Add `http response errors`.  Handle those errors.
* Add `LoginPage` and `RegistrationPage` - wrappers for `LoginForm` and `RegistrationForm` respectively.
* Handle login and registration requests in `Page`, rather then `Form` component.

### CODE REVIEW

**Comment by dkalinin** [comment]
Please merge approved validation PRs to master and merge master to api-server afterwards. Then I will review this PR.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

File: webclient/app/pages/not-found-page.js [hunk 1]
```diff
@@ -1,16 +1,28 @@
 import Component from '../components/component.js';
+import Services from '../services/services.js';
```
Comment (dkalinin) [inline webclient/app/pages/not-found-page.js:2]: Please remove.

File: webclient/app/app.js [hunk 2]
```diff
@@ -27,16 +27,27 @@ export default class Application extends Component {
    */
   initNestedComponents() {
     const {rootElement} = this;
-    const pageContainer = rootElement.querySelector('[data-type="application-box"]');
+    const {titleService} = Services;
```
Comment (dkalinin) [inline webclient/app/app.js:30]: - [ ] Let's remove Services singleton.

File: webclient/app/components/form/form.js [hunk 3]
```diff
@@ -99,7 +100,7 @@ export default class Form extends Component {
    * Hides error messages on the form input components.
    * @private
```
Comment (dkalinin) [inline webclient/app/components/form/form.js:101]: Please remove.

File: webclient/app/models/user-credentials.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  /**
+   * Instantiates UserCredentials value object.
+   *
+   * @param {string} login - user login.
+   * @param {string} password - password.
+   */
+  constructor(login, password) {
+    this.login = login;
+    this.password = password;
+    Object.freeze(this);
```
Comment (dkalinin) [inline webclient/app/models/user-credentials.js:29]: - [ ] Please remove.

File: webclient/app/pages/login-page.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ */
+export default class LoginPage extends AbstractFormPage {
+  /**
+   * Render @see {@link LoginForm} inside.
+   *
+   * @inheritdoc
+   */
+  initComponent() {
+    this.pageTitle = 'Authorization';
+    super.initComponent();
```
Comment (dkalinin) [inline webclient/app/pages/login-page.js:15]: - [ ] Please change sequence of these calls.

File: webclient/app/pages/abstract-form-page.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  /**
+   * Handler of successful response.
+   *
+   * @param {{token: string}} response - response containing token.
+   * @private
+   * @abstract
+   */
+  _resolveResponse(response) {
```
Comment (dkalinin) [inline webclient/app/pages/abstract-form-page.js:128]: - [ ] Please rename.

File: webclient/app/pages/abstract-form-page.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    super(container, {router, titleService, apiService});
+  }
+
+  /**
+   * @inheritdoc
+   * @return {string} - root element.
+   */
+  markup() {
+    return `
+      <div class="application-box" data-type="application-box" data-test="form-page-rendered"></div>
```
Comment (dkalinin) [inline webclient/app/pages/abstract-form-page.js:34]: form-page-rendered => form-page

File: webclient/tests/app/pages/login-page-tests.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    });
+
+    assert.verifySteps(['Page title is set: Authorization.'], 'Should properly set page title.');
+  });
+
+  test('should make log-in request.', (assert) => {
+    assert.expect(4);
+    const done = assert.async();
+
+    fixture.innerHTML = '';
```
Comment (dkalinin) [inline webclient/tests/app/pages/login-page-tests.js:48]: - [ ] Please remove.

File: webclient/tests/app/pages/login-page-tests.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    const router = {
+      redirectTo(hash) {
+        assert.step(`Should redirect when request succeeded. Hash: '${hash}'.`);
+
+        assert.verifySteps([
+          'Should set page title: Authorization.',
+          'Log-in request is made. Login: \'admin\'. Password: \'qwerty123A\'.',
+          'Should redirect when request succeeded. Hash: \'/registration\'.',
+        ], 'Should handle log-in request.');
+      },
```
Comment (dkalinin) [inline webclient/tests/app/pages/login-page-tests.js:73]: - [ ] Please move this to the end of the test.

File: webclient/tests/app/pages/login-page-tests.js (hunk 8 above)
Comment (YehorPytomets) [reply webclient/tests/app/pages/login-page-tests.js:48]: Fixed that by resetting link to `fixture` in `beforeEach` callback. `Qunit` [substitutes ](https://github.com/qunitjs/qunit/blob/89dcd90211fb129df48aa11b72dbafd9f5a18009/runner/fixture.js) fixture automatically before each test.

File: webclient/tests/app/pages/login-page-tests.js (hunk 9 above)
Comment (YehorPytomets) [reply webclient/tests/app/pages/login-page-tests.js:73]: I've tried different approaches, but I cannot move this to the end of the test. I think the reason is that I have asynchronous call to the `ApiService.logIn()`, which cannot be controlled by `await` inside the test. If I move it to the end of the test, I have my steps verified before actually calling them.
So, in order to check order of called `steps` I've left it inside `Router.redirectTo()`.
image

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

File: webclient/app/components/form/form.js [hunk 10]
```diff
@@ -136,4 +135,13 @@ export default class Form extends Component {
       }
     });
   }
+
+  /**
+   * Resolves authentication error.
+   *
+   * @param {AuthenticationError} error - authentication error.
+   */
+  resolveAuthenticationError(error) {
```
Comment (dkalinin) [inline webclient/app/components/form/form.js:144]: - [ ] Let's rename to show...

File: webclient/app/pages/login-page.js [hunk 11]
```diff
@@ -26,7 +26,7 @@ export default class LoginPage extends AbstractFormPage {
   /**
    * @inheritdoc
    */
-  _resolveResponse(response) {
+  _handleSuccessfulResponse(response) {
     const redirectOn = '/registration';
     this.properties.router.redirectTo(redirectOn);
```
Comment (dkalinin) [inline webclient/app/pages/login-page.js:31]: - [ ] Please move router out of pages.

File: webclient/tests/app/pages/login-page-tests.js (hunk 9 above)
Comment (YehorPytomets) [reply webclient/tests/app/pages/login-page-tests.js]: Left it as is.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

File: webclient/app/app.js [hunk 12]
```diff
@@ -27,23 +27,31 @@ export default class Application extends Component {
...
-        apiService,
-      }),
-      '/registration': () => new RegistrationPage(rootElement, {
-        router: this.router,
-        titleService,
-        apiService,
-      }),
+      [hashes.login]: (router) => new LoginPage(
+        rootElement, {
+          successfulResponseHandler: () => router.redirectTo(hashes.registration),
```
Comment (dkalinin) [inline webclient/app/app.js:39]: - [ ] Please rename.

**Pull request merged**